### PR TITLE
res_pjsip_path.c: Fix path when dialing using PJSIP_DIAL_CONTACTS()

### PR DESCRIPTION
**Pull request by IgorrG**
When using the PJSIP_DIAL_CONTACTS() function for use in the Dial()
command, the contacts are returned in text form, so the input to
the path_outgoing_request() function is a contact value of NULL.
The issue was reported in ASTERISK-28211, but was not actually fixed
in ASTERISK-30100. This fix brings back the code that was previously
removed and adds code to search for a contact to extract the path
value from it.

### CODE REVIEW

**Comment by IgorrG** [comment]
cherry-pick-to: 21
cherry-pick-to: 20
cherry-pick-to: 18

**Comment by gtjoseph** [comment]
Links to original issues if you're interested...
https://issues-archive.asterisk.org/ASTERISK-28211
https://issues-archive.asterisk.org/ASTERISK-30100

File: res/res_pjsip_path.c [hunk 1]
```diff
@@ -48,6 +48,109 @@ static struct ast_sip_aor *find_aor(struct ast_sip_contact *contact)
...
+{
+	struct ao2_iterator it_contacts;
+	struct ast_sip_contact *contact;
+	const pj_str_t *uri_username;
+	const pj_str_t *uri_hostname;
+	char *domain_name;
+	char *username;
+	RAII_VAR(struct ao2_container *, contacts, NULL, ao2_cleanup);
+
+	uri_hostname = ast_sip_pjsip_uri_get_hostname(uri);
```
Comment (mbradeen) [inline res/res_pjsip_path.c:128]: the uri is getting passed in and used to build these names, was that to compare against the contact when iterating?

**Review by mbradeen — CHANGES_REQUESTED** [review]

File: res/res_pjsip_path.c (hunk 1 above)
Comment (IgorrG) [reply res/res_pjsip_path.c:128]: Thanks for pointing this, I have figured out yesterday that some old patch uploaded. I'll update patch, correct way is to compare exact URI, without host/username extraction.

File: res/res_pjsip_path.c [hunk 2]
```diff
@@ -48,6 +48,105 @@ static struct ast_sip_aor *find_aor(struct ast_sip_contact *contact)
...
+	contact_buf[contact_buf_len] = '\0';
+
+	it_contacts = ao2_iterator_init(contacts, 0);
+	for (; (contact = ao2_iterator_next(&it_contacts)); ao2_ref(contact, -1)) {
+		if (!strcmp(contact_buf, contact->uri)) {
+			break;
+		}
+	}
+	ao2_iterator_destroy(&it_contacts);
+	return contact;
```
Comment (mbradeen) [inline res/res_pjsip_path.c:146]: if no match is found, what should be returned?

**Review by mbradeen — CHANGES_REQUESTED** [review]

File: res/res_pjsip_path.c (hunk 2 above)
Comment (IgorrG) [reply res/res_pjsip_path.c:146]: Got it, will fix

**Review by mbradeen — APPROVED** [review]

**Comment by gtjoseph** [comment]
I had to rebase this PR on current master so it would pick up the alembic fix that just merged.  When all the checks finish, this PR should merge successfully.

**Pull request merged**